aspnetcore/fundamentals/url-rewriting.md

```diff
@@ -0,0 +1,295 @@
+
+### Extension and Options
+Establish your URL rewrite and redirect rules by creating an instance of the `RewriteOptions` class with extension methods for each of your rules. Chain multiple rules in the order that you would like them processed. The `RewriteOptions` are passed into the URL Rewriting Middleware as it's added to the request pipeline with `app.UseRewriter(options);`.
+
+[!code-csharp[Main](url-rewriting/sample/Startup.cs?name=snippet1)]
+
+#### URL Redirect
+Use `AddRedirect()` to redirect requests. The first parameter will contain your regex for matching on the path of the incoming URL. The second parameter is the replacement string. The third parameter, if present, specifies the status code. If you don't specify the status code, it defaults to 302 (Found), which indicates that the resource has been temporarily moved or replaced.
+
+[!code-csharp[Main](url-rewriting/sample/Startup.cs?name=snippet1&highlight=2)]
+
+In a browser with developer tools enabled, make a request to the sample application with the URL **http://localhost:5000/redirect-rule/1234/5678**. The regex matches the request path on `redirect-rule/(.*)`, and the path is replaced with **redirected/1234/5678**. The redirect URL is sent back to the client with a 302 (Found) status code. The browser makes a new request at the redirect URL, which will appear in the browser's address bar. Since no rules in the sample application match on the redirect URL, the second request receives a 200 (OK) response from the application and the body of the response shows the redirect URL. A complete roundtrip is made to the server when a URL is *redirected*.
+
+Original Request: **http://localhost:5000/redirect-rule/1234/5678**
+
+![Browser window with Developer Tools tracking the requests and responses](url-rewriting/_static/add_redirect.png)
+
+The part of the expression contained by parentheses is called a *capture group*. The dot (`.`) of the expression means *match any character*, and the asterisk (`*`) signifies to *match the preceding character zero or more times*. Therefore, the last two path segments of the URL, `1234/5678`, are captured by capture group `(.*)`. Any value you provide in the request URL after **redirect-rule/** will be captured by this single capture group.
+
+In the replacement string, captured groups are injected into the string with the dollar sign (`$`) followed by the sequence number of the capture. The first capture group value is obtained with `$1`, the second with `$2`, and they continue in sequence for the capture groups in your regex. There is only one captured group in the redirect rule regex in the sample application, so there is only one injected group in the replacement string, which is `$1`. When the rule is applied, the URL becomes **http://localhost:5000/redirected/1234/5678**.
+
+#### URL Redirect to a secure endpoint
```

Comment:
I would also consider elevating the h4s to h2s, to get them into the internal TOC.  Logically they're under Extensions and Options, but I don't think elevating them would cause confusion.